controlplane/kubeadm/controllers/upgrade.go

```diff
@@ -82,33 +80,15 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, errors.Wrap(err, "failed to upgrade kubelet config map")
 	}
 
-	// If there is not already a Machine that is marked for upgrade, find one and mark it
-	selectedForUpgrade := requireUpgrade.Filter(machinefilters.HasAnnotationKey(controlplanev1.SelectedForUpgradeAnnotation))
-	if len(selectedForUpgrade) == 0 {
-		selectedMachine, err := r.selectAndMarkMachine(ctx, requireUpgrade, controlplanev1.SelectedForUpgradeAnnotation, controlPlane)
-		if err != nil {
-			logger.Error(err, "failed to select machine for upgrade")
-			return ctrl.Result{}, err
-		}
-		selectedForUpgrade = selectedForUpgrade.Insert(selectedMachine)
+	status, err := workloadCluster.ClusterStatus(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
 	}
 
-	replacementCreated := selectedForUpgrade.Filter(machinefilters.HasAnnotationKey(controlplanev1.UpgradeReplacementCreatedAnnotation))
-	if len(replacementCreated) == 0 {
-		// TODO: should we also add a check here to ensure that current machines not > kcp.spec.replicas+1?
-		// We haven't created a replacement machine for the cluster yet
-		// return here to avoid blocking while waiting for the new control plane Machine to come up
-		result, err := r.scaleUpControlPlane(ctx, cluster, kcp, ownedMachines, controlPlane)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		if err := r.markWithAnnotationKey(ctx, selectedForUpgrade.Oldest(), controlplanev1.UpgradeReplacementCreatedAnnotation); err != nil {
-			return ctrl.Result{}, err
-		}
-		return result, nil
+	if status.Nodes <= *kcp.Spec.Replicas {
```

Comment:
Having this to be on status.Nodes might be causing false negatives, for example during the time that takes for a machine to get up and running, the node isn't not going to be there (but the Machine will)

Reply:
Yeah, I mentioned that this logic is dependent on `TargetClusterControlPlaneIsHealthy` in scaleUp. That function ensures that each machine has a corresponding node, and short-circuits scaling up until that's true. So it ends up requeuing until the node appears or the machine disappears.

Reply:
I added a test for this too, just to try to ensure that this doesn't change https://github.com/kubernetes-sigs/cluster-api/pull/2887/files#diff-a9d99a2b1c85794cb5e38dd40e4d2722R85

Reply:
Can you add a comment to that effect around this if condition? Reading the code it's not immediately clear

Reply:
Added a comment and also changed the logic to only pass through the machines requiring upgrade.